File: files/300-399/313/CH-313.R
library(tidyverse)
library(readxl)

path = "files/300-399/313/CH-313 Table Transformation.xlsx"
input = read_excel(path, range = "B2:J8")
test  = read_excel(path, range = "L2:O7") %>% mutate(across(everything(), as.character))

shift4 = function(df){
  valid = !is.na(df) & str_trim(as.character(df)) != ""
  start = max.col(valid, ties.method = "first")
  start[!rowSums(valid)] = 1
  map_dfr(seq_len(nrow(df)), function(i) {
    v = unlist(df[i, ], use.names = FALSE)
    vals = v[start[i] + 0:3]; length(vals) = 4
    tibble(Date = vals[1], Product = vals[2], Customer = vals[3], Quantity = vals[4])
  })
}

output = shift4(input) %>%
  filter(row_number() > 1) %>%
  mutate(Date = janitor::excel_numeric_to_date(as.numeric(Date)) %>% as.character())

all.equal(output, test, check.attributes = FALSE)
# [1] TRUE

File: files/300-399/313/CH-313.py
import pandas as pd
from pprint import pprint

path = "300-399/313/CH-313 Table Transformation.xlsx"

input = pd.read_excel(path, usecols="B:J", skiprows=1, nrows=7)
test  = pd.read_excel(path, usecols="L:O", skiprows=1, nrows=5).astype(str)

output = []
for i, row in input.iterrows():
    first_valid_index = row.first_valid_index()
    last_index = row.last_valid_index()
    if first_valid_index is not None and last_index is not None:
        new_row = row.loc[first_valid_index:last_index]
        new_row.index = [f"Column{j+1}" for j in range(len(new_row))]
        output.append(new_row)

output = pd.DataFrame(output).reset_index(drop=True)
output.columns = output.iloc[0]
output = output[1:].reset_index(drop=True)